File: 1277. Count Square Submatrices with All Ones/Solution1277.py
from typing import List


class Solution1277:
    def countSquares(self, matrix: List[List[int]]) -> int:
        sum = 0
        r, c = len(matrix), len(matrix[0])
        dp = [[0]*(c+1) for _ in range(r+1)]
        for i in range(r):
            for j in range(c):
                if matrix[i][j] == 0:
                    dp[i+1][j+1] = 0
                else:
                    dp[i+1][j+1] = min(dp[i][j+1], dp[i+1][j], dp[i][j]) + 1
                    sum += dp[i+1][j+1]
        return sum

File: 1277. Count Square Submatrices with All Ones/Solution1277.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

class Solution1277 {
public:
    int countSquares(vector<vector<int>>& matrix) {
        int sum = 0;
        int row = matrix.size();
        int col = matrix[0].size();
        vector<vector<int>> dp(row+1, vector<int>(col+1, 0));
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (matrix[i][j] == 0) {
                    dp[i+1][j+1] = 0;
                } else {
                    dp[i+1][j+1] = min(min(dp[i][j+1], dp[i+1][j]), dp[i][j]) + 1;
                    sum += dp[i+1][j+1];
                }
                
            }
        }
        return sum;
    }
};